### PR TITLE
Layered store

### PR DESCRIPTION
**Pull request by icristescu**

### CODE REVIEW

**Comment by icristescu** [comment]
For the moment: 
- the `gc` is independent of the `move_to_next_layer` and it creates a copy of a repo using a `slice`, which as far as I understand keeps everything in memory. maybe not ideal? 
- the copy created by `gc` has a new name `root_"V"_version`  
- I modified the `Make_ext` functor in place, but maybe we want a different functor for this
- some minimal tests in `test/irmin-pack/layered.ml`

**Comment by icristescu** [comment]
In the current version, the layered store contains two stores under the hood: 
- the current RW store
- the "freezed" store

A `freeze` copies the contents of the current store into the "freezed" store and clears the current one. All operations following the freeze are done on a cleared store (this is probably not what we want, as I suspect Tezos would like to have an uninterrupted chain of commits.) 

`Make_layered` and `Make_ext_layered` are the functors for creating a layered store (which leads to some code duplications and to some refactoring of the irmin_pack Make functors). 

The layered store is implemented for `irmin-pack` and `irmin-mem` and passes some tests. There are some operations not implemented yet (documented in `src/irmin-test/layered_store.ml`) and also some performance fixes (the `copy` in `src/irmin/store.ml` could do less).

**Comment by icristescu** [comment]
I have written a small [doc](https://github.com/icristescu/irmin/blob/layers_doc/layers_doc/layered.md) explaining the PR.

**Comment by andreas** [comment]
Thanks for the doc! Quick question: Can the upper and lower layer use different storage, e.g. upper layer in memory, lower layer on disk?

**Comment by icristescu** [comment]
> Thanks for the doc! Quick question: Can the upper and lower layer use different storage, e.g. upper layer in memory, lower layer on disk?

Not for the moment, for now the upper and lower need to be of the same type of backend. But it's a nice idea, thanks ! I'll try to do it for a future improvement.

File: src/irmin/irmin.mli [hunk 1]
```diff
@@ -951,6 +951,9 @@ module type ATOMIC_WRITE_STORE = sig
   val close : t -> unit Lwt.t
   (** [close t] frees up all the resources associated to [t]. Any
       operations run on a closed store will raise {!Closed}. *)
+
+  val clear : t -> unit Lwt.t
```
Comment (CraigFe) [inline src/irmin/irmin.mli]: Is this (different from) / (better than) than just creating a new store and throwing away the old one?

File: src/irmin-mem/irmin_mem.mli [hunk 2]
```diff
@@ -33,5 +33,8 @@ module Atomic_write : Irmin.ATOMIC_WRITE_STORE_MAKER
 module Make : Irmin.S_MAKER
 (** An in-memory Irmin store. *)
 
+module Make_layered : Irmin.L_MAKER
```
Comment (CraigFe) [inline src/irmin-mem/irmin_mem.mli]: As discussed in person, it might be nice to avoid the proliferation of `Make_*` functors by expressing this as a functor over stores. I think it'll actually need to be a functor over the _existing_ `Make` functors (i.e. ∀ `Make` → `Make_layered`), to avoid the input store being too abstract to do anything useful with. Have you had a look at this?

File: src/irmin-test/common.ml [hunk 3]
```diff
@@ -14,6 +14,34 @@
...
+  let chars = "0123456789abcdefghijklmnopqrstABCDEFGHIJKLMNOPQRST-_." in
+  chars.[Random.int @@ String.length chars]
+
+let random_string n = String.init n (fun _i -> random_char ())
+
+let long_random_string = random_string (* 1024_000 *) 10
+
+let random_ascii_string n = String.init n (fun _i -> random_ascii ())
+
+let long_random_ascii_string = random_ascii_string 1024_000
```
[the remark this answers is not in the document]
Comment (icristescu) [reply src/irmin-test/common.ml]: Indeed, at some point, it would be nice to have several sizes of key and values for both irmin and index. And maybe use fuzzing to generate them.

File: src/irmin/irmin.mli (hunk 1 above)
Comment (icristescu) [reply src/irmin/irmin.mli]: `clear` is my next PR to extract from the layered store. 
I'm not sure what is best. Throwing away the old store means closing it, and leaving it to the user to remove all disk files generated by the store. Whereas `clear` handles it for the user.

Comment (CraigFe) [reply src/irmin/irmin.mli]: We could have a `close_and_clear` operation on the discarded store, if your concern is only user-responsibility for destroying the underlying resources. I brought it up because `clear` is that an imperative API (where I like things to be as functional as possible :slightly_smiling_face:), but looking more closely the rest of this signature is imperative anyway.

Comment (icristescu) [reply src/irmin/irmin.mli]: I think I prefer having a separate `clear` and `close`, instead of having two `close` operations. Alternatively, in Irmin-pack we have a fresh flag, for reopening a store and we could propagate it for all backend.

Comment (CraigFe) [reply src/irmin/irmin.mli]: Let's stick with `clear` and `close` then. I already dislike the `fresh` flag enough where it is :slightly_smiling_face:

File: src/irmin/layer.mli [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  include S.STORE
+
+  val freeze :
+    ?min:commit list -> ?max:commit list -> ?squash:bool -> repo -> unit Lwt.t
+  (** [freeze min max squash t] copies the upper layer of [t] into the lower
+      layer and clears the upper.
+
+      If [squash] then copy the commits, nodes and contents of the [max]
+      commits. Otherwise, copy the closure of commits, nodes and contents
+      between the commits [min] and [max]. *)
```
Comment (samoht) [inline src/irmin/layer.mli]: What happens if `max` is not set? What happens is `max` is not set AND `squash` is set? In these cases, would it make sense to look at the existing branches' HEAD in upper?

File: src/irmin/layer.ml [hunk 5]
```diff
@@ -0,0 +1,416 @@
+(*
```
Comment (samoht) [inline src/irmin/layer.ml]: It might make sense to add a new package (`irmin-layer`) for this new functionality.

File: src/irmin/layer.ml [hunk 6]
```diff
@@ -0,0 +1,416 @@
...
+ * Copyright (c) 2013-2019 Thomas Gazagnaire <thomas@gazagnaire.org>
```
Comment (samoht) [inline src/irmin/layer.ml]: wrong year :-)

File: src/irmin/layer.mli (hunk 4 above)
Comment (icristescu) [reply src/irmin/layer.mli]: The [current heads are used](https://github.com/mirage/irmin/pull/882/files#diff-e1c2498849d6289c1a35dec9cab695edR363) when max is empty. I added a line in the doc about it.

**Comment by icristescu** [comment]
Depends on #939 and #936

File: src/irmin-pack/layers_IO.ml [hunk 7]
```diff
@@ -77,3 +77,45 @@ module Unix : S = struct
...
+  val read_flip : Unix.t -> bool Lwt.t
+
+  val write_flip : bool -> Unix.t -> unit Lwt.t
+
+  val read_generation : Unix.t -> int64 Lwt.t
+
+  val write_generation : int64 -> Unix.t -> unit Lwt.t
+end
+
+module WrapUnix : WrapS = struct
```
Comment (samoht) [inline src/irmin-pack/layers_IO.ml:95]: As you've started doing, it might be interesting to go even further with this and use `Lwt_unix` functions here

File: src/irmin-pack/s.ml [hunk 8]
```diff
@@ -18,4 +18,177 @@ module type ATOMIC_WRITE_STORE = sig
...
+    LAYERED_CONTENT_ADDRESSABLE_STORE
+      with type key = key
+       and type value = V.t
+       and type index = index
+       and type U.index = index
+       and type L.index = index
+       and type U.key = key
+       and type L.key = key
+       and type U.value = V.t
+       and type L.value = V.t
```
Comment (CraigFe) [inline src/irmin-pack/s.ml:148]: re. the comment above, are they still multiple kinds of value if the types are constrained to be equal? I think I'm missing something...

File: src/irmin-test/layered_store.ml [hunk 9]
```diff
@@ -0,0 +1,958 @@
...
+      Alcotest.(check string) "x1" s "x1";
+      Lwt_list.iter_p
+        (fun i ->
+          check_layer_id repo
+            (S.Content_t (List.nth contents i))
+            ("layer id of x" ^ string_of_int i)
+            `Upper)
+        [ 0; 1; 3 ]
+      >>= fun () ->
+      Lwt.catch
```
Comment (CraigFe) [inline src/irmin-test/layered_store.ml:800]: [This helper function](https://github.com/mirage/irmin/blob/39454c5240adef463c1a37b7e0cf267827c1078a/test/irmin-pack/common.ml#L109) might be useful here and in a few other places.

File: src/irmin-test/layered_store.ml [hunk 10]
```diff
@@ -0,0 +1,958 @@
...
+    let p, r = Lwt.wait () in
+    let hook =
+      Hook.v (function
+        | `After_Clear ->
+            Log.debug (fun l -> l "test hook -- after_clear");
+            Lwt.async (fun () -> f () >|= fun () -> Lwt.wakeup r ());
+            Lwt.return_unit
+        | _ -> Lwt.return_unit)
+    in
+    (hook, p)
```
Comment (CraigFe) [inline src/irmin-test/layered_store.ml:895]: These look like they could be cut down with a helper function :slightly_smiling_face:

**Review by CraigFe — COMMENTED** [review]
Looking good :slightly_smiling_face: Here's an initial round of questions / comments after a fairly cursory glance through. 

The `iter` function looks like it be added separately; did we have a PR for that at some point?

File: src/irmin-layers/irmin_layers_intf.ml [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+  (** [layer_id t store_handle] returns the layer where an object, identified by
+      its hash, is stored. *)
+
+  val async_freeze : unit -> bool
+  (** [async_freeze t] returns true if there is an ongoing freeze. To be used
+      with caution, as a freeze can start (or stop) just after the test. It is
+      helpful when a single freeze is called, to check whether it completed or
+      not. *)
+
+  val upper_in_use : repo -> string
```
[the remark this answers is not in the document]
Comment (icristescu) [reply src/irmin-layers/irmin_layers_intf.ml]: I only use it in a test, I moved it to the private functor.

File: src/irmin-pack/irmin_pack_layers.ml [hunk 12]
```diff
@@ -0,0 +1,765 @@
...
+        | Commit_t k -> Commit.CA.layer_id t.commit k
+        | Node_t k -> Node.CA.layer_id t.node k
+        | Content_t k -> Contents.CA.layer_id t.contents k
+
+      let clear_previous_upper t =
+        Log.debug (fun l -> l "clear previous upper");
+        Contents.CA.clear_previous_upper t.contents >>= fun () ->
+        Node.CA.clear_caches_next_upper t.node;
+        Commit.CA.clear_caches_next_upper t.commit;
+        Branch.clear_previous_upper t.branch
```
[the remark this answers is not in the document]
Comment (icristescu) [reply src/irmin-pack/irmin_pack_layers.ml:454]: Hm, that would be nice, but I don't know how to do that. Can you point me to an example?

File: src/irmin-pack/irmin_pack_layers.ml [hunk 13]
```diff
@@ -0,0 +1,765 @@
...
+
+      let upper_in_use t =
+        if t.flip then upper_root1 t.config else upper_root0 t.config
+
+      let offset t = Contents.CA.offset t.contents
+
+      let rec copy_newies_to_next_upper t former_offset =
+        let offset = offset t in
+        if offset -- former_offset >= 64L then
+          Contents.CA.copy_newies_to_next_upper t.contents >>= fun () ->
```
[the remark this answers is not in the document]
Comment (icristescu) [reply src/irmin-pack/irmin_pack_layers.ml:477]: After a call to freeze we keep using the same upper as before, to ensure that adding objects after a freeze started is as fast as before. Towards the end of the freeze, they are copied to the next upper. And it is only at the end of the freeze that we flip uppers.
I called "newies" all objects added to the current upper once a freeze started. I wanted to have a somewhat suggestive name for these functions, but I'm not sure it is :) also added a few comments to clarify this.

File: src/irmin-pack/irmin_pack_layers.ml [hunk 14]
```diff
@@ -0,0 +1,765 @@
...
+          X.Repo.clear_previous_upper t)
+      >>= fun () ->
+      (* RO reads generation from pack file to detect a flip change, so it's
+         ok to write the flip file outside the lock *)
+      X.Repo.write_flip t >>= fun () ->
+      Lwt_mutex.unlock freeze_lock;
+      Log.debug (fun l -> l "free lock");
+      may (fun f -> f `After_Clear) hook
+    in
+    Lwt.async (fun () ->
```
[the remark this answers is not in the document]
Comment (icristescu) [reply src/irmin-pack/irmin_pack_layers.ml:702]: Indeed, but maybe we can do this in a later PR?

[the remark this answers is not in the document]
Comment (CraigFe) [reply src/irmin-pack/irmin_pack_layers.ml:702]: Sure :+1:

File: src/irmin-pack/irmin_pack_layers.ml (hunk 12 above)
[the remark this answers is not in the document]
Comment (CraigFe) [reply src/irmin-pack/irmin_pack_layers.ml:454]: It would first be necessary to define the common supertype of the four modules as its own module type (since `Branch` is an atomic write store while the others are not) – that may already exist. Then you can do something like:

```ocaml
let iter_lwt f { contents; node; commit; branch } =
  f (module Contents.CA) contents >>= fun () ->
  f (module Node.CA) node >>= fun () ->
  f (module Commit.CA) commit >>= fun () ->
  f (module Branch) branch

let dispatch t = function
 | Commit_t k -> f (module Commit.CA) t.commit k
 | Node_t k -> f (module Node.CA) t.node k
 | Content_t k -> f (module Contents.CA) t.contents k
```

etc. etc.

[the remark this answers is not in the document]
Comment (icristescu) [reply src/irmin-pack/irmin_pack_layers.ml:495]: Do you mind having a look [here](https://github.com/mirage/irmin/pull/882/commits/6804bf94b26f6fbcb83981e2f11ab977b46947d8#diff-52ce06d08638434eac31d8f0240b3b5eR286)? 

I can't figure out how to add `f` as an argument to `let iter_lwt f t =...` : I have to write its type explicitly, so that the type of the store is derived locally at each application of `f`.

[the remark this answers is not in the document]
Comment (CraigFe) [reply src/irmin-pack/irmin_pack_layers.ml:495]: Hmm, yes. Sorry, my suggestion was insufficient. `f` needs a forall-quantifier, which then requires defining a record type:

```ocaml
        type 'a store_fn = {
          f : 't. (module S.STORE with type t = 't) -> 't -> 'a;
        }
        [@@ocaml.unboxed]

        let iter_lwt (f : unit Lwt.t store_fn) t : unit Lwt.t =
          f.f (module Contents) t.contents >>= fun () ->
          f.f (module Nodes) t.node >>= fun () ->
          f.f (module Commits) t.commit >>= fun () ->
          f.f (module Branch) t.branch
```

Sorry for wasting your time :slightly_smiling_face:

[the remark this answers is not in the document]
Comment (icristescu) [reply src/irmin-pack/irmin_pack_layers.ml:501]: That's neat! Thanks for helping me out :)

**Review by pascutto — APPROVED** [review]
Let's merge this, so we can make small individual adjustments on `master` instead of tweaking this PR!

File: src/irmin-pack/s.ml [hunk 15]
```diff
@@ -21,3 +21,188 @@ module type ATOMIC_WRITE_STORE = sig
...
+end
+
+module type LAYERED_INODE = sig
```
Comment (CraigFe) [inline src/irmin-pack/s.ml:151]: I think there's quite a bit of duplication between the various `LAYERED_` signatures that would be good to refactor at some point.

File: test/irmin-pack/data/empty_store/index/lock [hunk 16]
```diff
@@ -0,0 +1 @@
+78919
```
Comment (CraigFe) [inline test/irmin-pack/data/empty_store/index/lock:1]: Did you mean to commit this?

File: test/irmin-pack/concurrent.ml [hunk 17]
```diff
@@ -0,0 +1,238 @@
+open Lwt.Infix
```
Comment (CraigFe) [inline test/irmin-pack/concurrent.ml:1]: I think this test can be considered separate from the layered store and split to a separate PR. I have some concerns as it stands.

File: test/irmin-pack/migration.ml [hunk 18]
```diff
@@ -36,111 +27,201 @@ let goto_project_root () =
...
-        m
-          "Checking old bindings are still reachable post-migration (cached RO \
-           instance)");
-    let* r = S.Repo.v conf_ro in
-    check_repo r archive
-  in
-  let* new_commit =
-    let* rw = S.Repo.v conf_rw in
-    Log.app (fun m -> m "Checking new commits can be added to the V2 store");
+module type Migrate_Store = sig
```
Comment (CraigFe) [inline test/irmin-pack/migration.ml:36]:
```suggestion
module type Migrate_store = sig
```

File: test/irmin-pack/migration.ml [hunk 19]
```diff
@@ -36,111 +27,201 @@ let goto_project_root () =
...
     let* () = check_commit ro new_commit [ ([ "c" ], "x") ] in
     S.Repo.close ro
-  in
-  Lwt.return_unit
+
+  let v1_to_v2 =
+    v1_to_v2 ~uncached_instance_check_idempotent ~uncached_instance_check_commit
+end
+
+module Config_Layered_Store = struct
```
Comment (CraigFe) [inline test/irmin-pack/migration.ml:175]:
```suggestion
module Config_layered_store = struct
```

File: src/irmin-pack/inode.ml [hunk 20]
```diff
@@ -769,42 +797,29 @@ struct
...
 
   let close = Inode.close
 
-  let sync = Inode.sync
-
   let clear = Inode.clear
 
   let clear_caches = Inode.clear_caches
+
+  let add _ _ = failwith "not implemented"
```
[the remark this answers is not in the document]
Comment (icristescu) [reply src/irmin-pack/inode.ml:820]: We cannot not implement them here, as the [`Inode.Make_ext`](https://github.com/mirage/irmin/blob/master/src/irmin-pack/inode_intf.ml#L83) functor is creating an `Irmin.CONTENT_ADDRESSABLE_STORE`. Both inodes and inodes-layers override this function in their `Make` functor, but they use different versions of the `unsafe_append` from `pack`. So we cannot provide with a default/common implementation for `add`.

File: src/irmin-pack/inode_layers.ml [hunk 21]
```diff
@@ -0,0 +1,158 @@
...
+  let add t v = save t v.Val.v >|= fun () -> hash v
+
+  let unsafe_add t k v =
+    check_hash k (hash v);
+    save t v.Val.v
+
+  let v = Inode.v
+
+  let sync ?on_generation_change:_ _ =
+    Fmt.failwith "not implemented for the node store"
```
[the remark this answers is not in the document]
Comment (icristescu) [reply src/irmin-pack/inode_layers.ml]: It's because inodes inherits pack signature. However the sync from inodes-layers (or inodes) [is not used](https://github.com/mirage/irmin/blob/master/src/irmin-pack/irmin_pack.ml#L234). But I modified it in inodes-layers to have a default implementation and so to be similar to the sync in inodes.

File: src/irmin-pack/pack.ml [hunk 22]
```diff
@@ -269,10 +272,12 @@ struct
           else Tbl.add t.staging k v;
           Lru.add t.lru k v
 
+    let return = Lwt.pause
```
[the remark this answers is not in the document]
Comment (icristescu) [reply src/irmin-pack/pack.ml]: I renamed it to `pause` and hopefully it's more explicit this way. But I would like to keep this function, because we might have to replace with `Lwt_main.yield` or `auto_yield` if the freeze thread blocks again, and its easier to do it this way.

File: src/irmin-pack/layered.mli [hunk 23]
```diff
@@ -0,0 +1,37 @@
...
+module Content_addressable
+    (H : Irmin.Hash.S)
+    (Index : Pack_index.S)
+    (Pack : Pack.S with type index = Index.t and type key = H.t) :
+  S.LAYERED_CONTENT_ADDRESSABLE_STORE
+    with type key = Pack.key
+     and type value = Pack.value
+     and type index = Pack.index
+     and module U = Pack
+     and module L = Pack
```
[the remark this answers is not in the document]
Comment (icristescu) [reply src/irmin-pack/layered.mli]: I think having `U` and `L` in the code makes a bit more readable.

File: src/irmin-pack/s.ml [hunk 24]
```diff
@@ -21,3 +21,188 @@ module type ATOMIC_WRITE_STORE = sig
...
+  include Pack.S
+
+  module U : Pack.S
+
+  module L : Pack.S
+
+  val v :
+    [ `Read ] U.t ->
+    [ `Read ] U.t ->
+    [ `Read ] L.t option ->
```
[the remark this answers is not in the document]
Comment (icristescu) [reply src/irmin-pack/s.ml:35]: It's because I'm using the function `val v : Conf.t -> [ `Read ] t Lwt.t` (https://github.com/mirage/irmin/blob/master/src/irmin/s.ml) to build the stores from which I then build a layered store.

File: src/irmin-pack/s.ml (hunk 15 above)
Comment (icristescu) [reply src/irmin-pack/s.ml:151]: Agree!

File: src/irmin-pack/inode.ml (hunk 20 above)
[the remark this answers is not in the document]
Comment (CraigFe) [reply src/irmin-pack/inode.ml:820]: Sure, my question was unclear, but I meant something like "why define `add` at all if we cannot implement it?". We could let the `inodes` and `inodes-layers` functors provide their own definitions, without needing to have a stub here.

File: src/irmin-pack/pack.ml (hunk 22 above)
[the remark this answers is not in the document]
Comment (CraigFe) [reply src/irmin-pack/pack.ml]: Sure, sounds good. Note that both `yield` and `auto_yield` are unix-only, so we probably cannot rely on them long-term.

File: src/irmin-pack/layered.mli (hunk 23 above)
[the remark this answers is not in the document]
Comment (CraigFe) [reply src/irmin-pack/layered.mli]: Sure, I can see that. However, it's probably only true if the code doesn't apply `U` and `L` to the wrong stores, however, which is not enforced by the type system. Is it necessary to have them be equal, or is this done to avoid a functor parameter?

File: src/irmin-pack/s.ml (hunk 24 above)
[the remark this answers is not in the document]
Comment (CraigFe) [reply src/irmin-pack/s.ml:35]: Hmm, that type looks broken to me too, but it's not related to this PR.

File: src/irmin-pack/inode.ml (hunk 20 above)
[the remark this answers is not in the document]
Comment (icristescu) [reply src/irmin-pack/inode.ml]: There might be another way, but to remove the stub `add` I had to specialise the `Make_ext` functor to inodes. I couldn't figure a way to reuse the `Make_ext` of inodes in inode-layers, so I duplicated the code needed in inodes. It's not so much code, I think it is a good compromise for now.

File: src/irmin-pack/layered.mli (hunk 23 above)
[the remark this answers is not in the document]
Comment (icristescu) [reply src/irmin-pack/layered.mli]: I removed the module equality and give `U` and `L` as functor parameters, and the type system does not let me write `U.function t.lower` anymore. Thanks!

[the remark this answers is not in the document]
Comment (CraigFe) [reply src/irmin-pack/layered.mli]: Thanks, I prefer this :slightly_smiling_face:

File: src/irmin-pack/inode.ml (hunk 20 above)
[the remark this answers is not in the document]
Comment (CraigFe) [reply src/irmin-pack/inode.ml]: Sure; let's leave it for now.

**Review by CraigFe — APPROVED** [review]
